### PR TITLE
[MERGE] crm: Migrate form/kanban views to owl

### PR DESCRIPTION
**Pull request by tivisse**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/crm_iap_mine/static/src/views/generate_leads_views.xml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+<templates>
+    <t t-name="LeadMiningRequest.generate_leads_button" owl="1">
+        <button t-if="isSalesManager" type="button" class="btn btn-secondary o_button_generate_leads" t-on-click="onClickGenerateLead">
+            Generate Leads
+        </button>
+    </t>
+
+    <t t-name="LeadMiningRequestListView.buttons" t-inherit="web.ListView.Buttons" t-inherit-mode="primary" owl="1">
+        <!-- Before the export button -->
+        <xpath expr="//t[@t-if='props.showButtons']//t[3]" position="before">
```
Comment (awa-odoo) [inline addons/crm_iap_mine/static/src/views/generate_leads_views.xml:12]: Is this targeting the "t" element based on position?
Does not seem super robust, can't we use a class or a name or something?

File: addons/crm_iap_mine/views/crm_lead_views.xml [hunk 2]
```diff
@@ -12,7 +12,7 @@
         </field>
     </record>
 
-    <record id="crm_lead_view_kanban_opportunity" model="ir.ui.view">
+    <!-- <record id="crm_lead_view_kanban_opportunity" model="ir.ui.view">
```
Comment (awa-odoo) [inline addons/crm_iap_mine/views/crm_lead_views.xml:15]: Commented code? :grimacing:

File: addons/web/static/src/views/kanban/kanban_animated_number.xml [hunk 3]
```diff
@@ -3,6 +3,7 @@
 
     <t t-name="web.KanbanAnimatedNumber" owl="1">
         <div class="o_kanban_counter_side ms-2 text-900 text-nowrap" t-att-class="state.value !== props.value and props.animationClass" t-att-title="props.title">
+            <t t-slot="prefix"/>
```
Comment (awa-odoo) [inline addons/web/static/src/views/kanban/kanban_animated_number.xml:6]: (Curious, what happens if this is undefined? Property handled by owl?)

File: addons/web/static/src/views/kanban/kanban_arch_parser.js [hunk 4]
```diff
@@ -104,13 +104,7 @@ export class KanbanArchParser extends XMLParser {
...
         const progressBar = xmlDoc.querySelector("progressbar");
         if (progressBar) {
-            const attrs = extractAttributes(progressBar, ["field", "colors", "sum_field", "help"]);
-            progressAttributes = {
-                fieldName: attrs.field,
-                colors: JSON.parse(attrs.colors),
-                sumField: fields[attrs.sum_field] || false,
-                help: attrs.help,
-            };
+            progressAttributes = this.parseProgressBar(progressBar, fields);
```
Comment (awa-odoo) [inline addons/web/static/src/views/kanban/kanban_arch_parser.js:107]: Shouldn't these changes in "web" be placed in a separate commit?

File: addons/web/static/src/views/kanban/kanban_model.js [hunk 5]
```diff
@@ -270,7 +270,16 @@ class KanbanGroup extends Group {
     updateAggregates(groupData) {
         const fname = this.groupByField.name;
         const { sumField } = this.model.progressAttributes;
-        const group = groupData.find((g) => this.valueEquals(g[fname]));
+        const group = groupData.find((g) => {
```
Comment (awa-odoo) [inline addons/web/static/src/views/kanban/kanban_model.js:273]: It feels as if these changes are related to both the MRR feature and the forecast feature.

Maybe we should split into different commits to explain a bit more the changes introduced?

**Review by awa-odoo — COMMENTED** [review]
Hello @tivisse @Williambraecky 

First of all, thank you for your work on this, it was not an easy port considering there was a lot of code and custom behaviors involved!

I left a few comments / suggestions / questions.

Did not look at the forecast part as I don't really know how it works, but maybe @abd-msyukyu-odoo will want to take a look if he has some time?

Cheers!

File: addons/crm/static/src/views/crm_form/crm_form.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+/**
+ * This Form Controller makes sure we display a rainbowman message
+ * when the stage is won, even when we click on the statusbar.
+ * When the stage of a lead is changed and data are saved, we check
+ * if the lead is won and if a message should be displayed to the user
+ * with a rainbowman like when the user click on the button "Mark Won".
+ */
+
+export class CrmFormField extends Field {
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/crm/static/src/views/crm_form/crm_form.js:18]: Indeed this could be a widget 😅 
Though it will need a little bit of chippotage since form views are still kinda weird 👍

File: addons/crm/static/src/views/crm_form/crm_form.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+     * message.
+     *
+     * This method will also simulate a real "force_save" on the email and phone
+     * when needed. The "force_save" attribute only works on readonly field. For our
+     * use case, we need to write the email and the phone even if the user didn't
+     * change them, to synchronize those values with the partner (so the email / phone
+     * inverse method can be called).
+     *
+     * We base this synchronization on the value of "partner_phone_update"
+     * and "partner_phone_update", which are computed fields that hold a value
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/crm/static/src/views/crm_form/crm_form.js:76]: I guess it was already like that, I changed it to `partner_email_update`

File: addons/crm/static/src/views/crm_kanban/crm_kanban_model.js [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        const sourceGroup = this.groups.find((g) => g.id === dataGroupId);
+        const targetGroup = this.groups.find((g) => g.id === targetGroupId);
+        if (
+            dataGroupId !== targetGroupId &&
+            sourceGroup &&
+            targetGroup &&
+            sourceGroup.groupByField.name === "stage_id"
+        ) {
+            const record = targetGroup.list.records.find((r) => r.id === dataRecordId);
+            await checkRainbowmanMessage(this.model.ormService, this.model.effect, record.resId);
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/crm/static/src/views/crm_kanban/crm_kanban_model.js:40]: `dataRecordId` is a unique id in form of a string like `datapoint_39` so no :/

File: addons/crm/static/src/views/crm_kanban/crm_kanban_model.js [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        }
+        return res;
+    }
+}
+
+export class CrmKanbanGroup extends CrmKanbanModel.Group {
+    updateAggregates(groupData) {
+        super.updateAggregates(...arguments);
+        const rrField = this.model.progressAttributes.recurring_revenue_sum_field;
+        const group = groupData.find(g => this.valueEquals(g[this.groupByField.name]));
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/crm/static/src/views/crm_kanban/crm_kanban_model.js:50]: `groupByField` is the current group by indeed.
It's a singular group as is `this`, basically `groupData` is the result of the `web_search_group`, this method is called whenever groups have to be reloaded to update some data, for instance if you have a progressbar it will update the `sum_field`; here we do the same but for the recurring revenue field :)

File: addons/crm/static/tests/crm_kanban_progress_bar_mrr_sum_field_tests.js [hunk 10]
```diff
@@ -161,36 +150,42 @@ QUnit.module('Crm Kanban Progressbar', {
...
+        reccurringRevenueNoValues = [...target.querySelectorAll('.o_kanban_counter_side:nth-child(3)')].map((elem) => elem.textContent);
+        assert.deepEqual(reccurringRevenueNoValues, ['+25', '+40'],
+        "counter should display the sum of recurring_revenue_monthly correctly after drag and drop");
+        
         //Activate "planned" filter on first column
-        await testUtils.dom.click(kanban.el.querySelector('.o_kanban_group:nth-child(2) .progress-bar[data-filter="planned"]'));
+        await click(target.querySelector('.o_kanban_group:nth-child(2) .progress-bar[aria-valuenow="2"]'), null);
+        // TODO: Owl bug: kanban animated numbers are rendered before being loaded
+        // kanban.model.load();
+        // await nextTick();
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/crm/static/tests/crm_kanban_progress_bar_mrr_sum_field_tests.js:184]: In this case it is not necessary. Forgot to remove.
The one above will hopefully be fixed by: https://github.com/odoo/odoo/pull/99748

File: addons/crm/static/tests/crm_rainbowman_tests.js [hunk 11]
```diff
@@ -65,320 +74,275 @@ odoo.define('crm.form_rainbowman_tests', function (require) {
...
-                    assert.step(result || "no rainbowman");
-                }
-                return result;
-            };
-            const kanban = await createView(this.testKanbanView);
+    QUnit.test("team record 30 days, click on statusbar", async function (assert) {
+        assert.expect(2);
+
+        this.testFormView.resId = 2;
+        this.testFormView.mockRPC = async function (route, args, performRpc) {
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/crm/static/tests/crm_rainbowman_tests.js:155]: Sure good idea

File: addons/crm/static/tests/tours/crm_forecast_tour.js [hunk 12]
```diff
@@ -56,29 +56,32 @@ tour.register('crm_forecast', {
     }, {
         trigger: ".o_kanban_record .o_kanban_record_title:contains('Test Opportunity 1')",
         content: "move to the next month",
-        run: function (actions) {
-            const undefined_groups = $('.o_column_title:contains("None")').length;
-            actions.drag_and_drop(` .o_opportunity_kanban .o_kanban_group:eq(${1 + undefined_groups})`, this.$anchor);
-        }
+        run: function () {
+            // OWL BUG: current implementation of drag_and_drop does not work with owl views
+            const undefined_groups = $('.o_column_title:contains("false")').length;
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/crm/static/tests/tours/crm_forecast_tour.js:61]: Ideally yes 😄

File: addons/crm/static/tests/tours/crm_rainbowman.js [hunk 13]
```diff
@@ -14,20 +14,37 @@
...
             trigger: "button.o_kanban_add",
             content: "create lead",
+        }, { //OWL BUG: can not move a newly created record directly, go to lead's form view and back.
+            trigger: ".o_kanban_record .o_kanban_record_title:contains('Test Lead 1')",
+            content: "open lead",
+        }, {
+            trigger: '.o_back_button',
+            content: 'navigate back to the kanban view',
+            position: "bottom",
+            run: "click",
```
[the remark this answers is not in the document]
Comment (Williambraecky) [reply addons/crm/static/tests/tours/crm_rainbowman.js:34]: It's as you prefer, I was just making the branch green, I already reported the bug and hopefully it can be fixed soon :)
I was told this PR was blocking some other stuff.

File: addons/crm_iap_mine/static/src/views/generate_leads_views.xml (hunk 1 above)
Comment (Williambraecky) [reply addons/crm_iap_mine/static/src/views/generate_leads_views.xml:12]: Our options are quite limited unfortunately.. We can either:
 - `<t t-elif="activeActions['create'] and (!model.root.isGrouped or !editable)">` position="after"
 - `<t t-if="nbTotal and !nbSelected and activeActions['exportXlsx'] and isExportEnable and !env.isSmall">` position="before"
Perhaps something like this is more appropriate:
```suggestion
        <xpath expr="//t[contains(@t-if, 'isExportEnable')]" position="before">
```

File: addons/web/static/src/views/kanban/kanban_arch_parser.js (hunk 4 above)
Comment (Williambraecky) [reply addons/web/static/src/views/kanban/kanban_arch_parser.js:107]: Sure

File: addons/web/static/src/views/kanban/kanban_model.js (hunk 5 above)
Comment (Williambraecky) [reply addons/web/static/src/views/kanban/kanban_model.js:273]: This is actually a bugfix, it should be in a separate commit as well.

File: addons/crm/static/src/js/tours/crm.js [hunk 14]
```diff
@@ -39,11 +39,11 @@ tour.register('crm_tour', {
     content: Markup(_t("Now, <b>add your Opportunity</b> to your Pipeline.")),
     position: "bottom",
 }, {
-    trigger: ".o_opportunity_kanban .o_kanban_group:first-child .o_kanban_record:last-child .oe_kanban_content",
+    trigger: ".o_opportunity_kanban .o_kanban_group:first-child .o_kanban_record:last-of-type .oe_kanban_content",
```
Comment (Williambraecky) [inline addons/crm/static/src/js/tours/crm.js:42]: About this change: `last-child` only works if it is the actual _last_ child, there were some changes to the dom and now after the kanban_records there is a "button" to create new records as well.

**Comment by tivisse** [comment]
@robodoo r+ merge

**Comment by tivisse** [comment]
@robodoo rebase-ff
